### PR TITLE
Add SIGHUP handler to pbench indexer

### LINKED ISSUE #2060: Add SIGHUP signal handler to the pbench indexer
SIGHUP: re-evaluate the list of tar balls to index after indexing of the current tar ball has finished

- Report the current state of indexing
            - SIGHUP
            - Current tar ball being processed
            - How many remaining
            - How many have been completed
            - Errors encountered
            - Handler Behavior:
            1. Kicks a special thread to log indexing state data
            2. No exception raised

### PR DESCRIPTION
**Pull request by riya-17**
Fixes #2060

Re-evaluate the list of tar balls to index after
indexing of the current tar ball has finished.

Report the current state of indexing
- SIGHUP
- Current tar ball being processed
- How many remaining
- How many have been completed

### CODE REVIEW

File: lib/pbench/server/indexing_tarballs.py
Comment (portante) [inline lib/pbench/server/indexing_tarballs.py:53]: Can you isolate these changes to add the `archive` and `qdir` parameters in a separate commit from the signal handling changes?

**Review by portante — CHANGES_REQUESTED** [review]
Good start.

Looks like you need to resolve a "LGTM.com" check that failed.

Also, the original intention of SIGHUP was to simply report the state of indexing.  While I like the idea of having SIGHUP also cause the indexer to re-evaluate the list of things to index, the goal was to have it immediately report the state of the indexing via log output while the indexing was on-going.

To do that, we'd need the interrupt handling to "kick" a thread which would perform the log output based on known state.

File: lib/pbench/server/indexing_tarballs.py
Comment (riya-17) [reply lib/pbench/server/indexing_tarballs.py:53]: Sure and will work on the reviiew

Comment (dbutenhof) [reply lib/pbench/server/indexing_tarballs.py:53]: Also, add SIGHUP to the pbench-index `main` docstring with the other exceptions!

Comment (riya-17) [reply lib/pbench/server/indexing_tarballs.py:53]: ok

File: lib/pbench/server/indexing_tarballs.py
Comment (dbutenhof) [inline lib/pbench/server/indexing_tarballs.py:442]: "recollecting" sounds odd here, and the normal sense of the word doesn't fit. How about "refreshing list of tarballs to index"??

Also, we continue on despite the failure... but on an error `collect_tb()` returns an empty tarball list, so you've forgotten whatever we started with.

I think you should try `status, new_tb = self.collect_tb()` instead, and assign `tarballs = new_tb` only if it succeeded.

Also, how does the iteration over `tarballs` work when you're completely replacing the list with a new list? Does Python handle this the way we want? (I don't know offhand.)

Finally, you're reporting how many we've done and how many remain based on "idx" relative to the number of elements in `tarballs` ... but `collect_tb()` won't find the datasets we've already indexed in the loop, so `tarballs` will be the number _remaining_ from the original list (unless more have been added). So, for example, if we find 10, index 5, and then get a `SIGHUP`, our `idx` is 5, and `collect_tb()` returns the 5 we haven't indexed yet... so your message will report that we've indexed 5 and there are 0 remaining (5 - 5). I think this logic needs to be re-worked.

While the "update the list" is an intriguing idea, remember that the indexer will be triggered again in a short time, so I'm not sure how useful that really is. The consistency issues would go away if you dropped that and just continued to iterate through the original list.

Eventually (and hopefully not that far off) the indexer will be demand-driven by messages that a new dataset is available.

File: server/bin/pbench-index.py
Comment (dbutenhof) [inline server/bin/pbench-index.py:115]: Reporting errors encountered might be useful, but you're not doing that.

File: server/bin/pbench-index.py
Comment (dbutenhof) [inline server/bin/pbench-index.py:117]: The output isn't being done in a thread; you're generating it inline. Which I think is fine, since this isn't a lot of code. Although I think the original idea was that instead of synchronously looking for an interrupt the next time around the loop at this one spot, the interrupt handler would create a thread that would log the immediate status in parallel, regardless of where in the loop we are. (And note that trying to refresh `tarballs` in that context would be even more problematic than it is here.)

**Review by dbutenhof — CHANGES_REQUESTED** [review]
I'm not sure this will even work as written... replacing a list you're iterating over is a potentially serious problem. There are also some inconsistencies in behavior and descriptions.

File: server/bin/pbench-index.py
Comment (webbnh) [inline server/bin/pbench-index.py:186]: Reassigning `tarballs` here may not do what you hope (see the "Note" at the end of [this section](https://docs.python.org/3/reference/compound_stmts.html#the-for-statement)).

If you really want to implement this functionality, you should probably use a "queue" model, where you start with a list of tarballs and destroy the list as you process it (e.g., use a `while tarballs` instead of the `for`, and remove the tarball from the list before processing it).

**Review by webbnh — APPROVED** [review]
These changes look OK to me, but beware of the issue I noted.

File: server/bin/pbench-index.py
Comment (webbnh) [reply server/bin/pbench-index.py:117]: I think doing it "in-line" is far better than trying to use a thread:

- Python threads don't actually offer much concurrency -- Python bytecode execution is done under a global lock, so a thread executing "Python code" only executes in parallel with other threads which _aren't_ running "Python code" (like stuff doing I/O or executing an external library such as NumPy which is written in C).
- Signal delivery is apparently also synchronous with respect to bytecode execution, so the notification is going to wait until "a good time", anyway.  (I.e., it might not actually interrupt anything.)
- The in-line approach is _much_ simpler and easier to understand and maintain.
- The execution of the interrupt is not performance-critical.
- Adding threads makes the update of the `tarballs` list much more complicated.

File: server/bin/pbench-index.py
Comment (webbnh) [inline server/bin/pbench-index.py:108]: Looks like there should be a blank line before this section, as there are for the other sections.

File: lib/pbench/server/indexing_tarballs.py
Comment (riya-17) [reply lib/pbench/server/indexing_tarballs.py:442]: > "recollecting" sounds odd here, and the normal sense of the word doesn't fit. How about "refreshing list of tarballs to index"??

I will change it to this or webb's  suggestion is also a good one



> Also, we continue on despite the failure... but on an error `collect_tb()` returns an empty tarball list, so you've forgotten whatever we started with.

I thought outside for loop will take care of that by completing the iterations, breaking it here might create unwanted terminations



> I think you should try `status, new_tb = self.collect_tb()` instead, and assign `tarballs = new_tb` only if it succeeded.

yeah sure I thought reassigning directly will make more sense



> Also, how does the iteration over `tarballs` work when you're completely replacing the list with a new list? Does Python handle this the way we want? (I don't know offhand.)

when the signal for the SIGHUP handler is called then it completes processing of the current tarball and then checks whether the signal is interrupted. Unindexed tarballs are all present in the "To-Index" folder so it will contain all the new tarballs that just came in and also all the tarballs that would get processed after this iteration. So, if we re-check that folder then it would contain all the tarballs that need to be indexed else they will get duplicated. like we are processing [1, 2, 3, 4] and [1, 2] are indexed, "To-index" folder will contain [3, 4]  as well as new tarballs that came in [5, 6]. So, we can just check tarballs in the "to-index" folder and replace it with an existing list which is 'tarballs'.



> Finally, you're reporting how many we've done and how many remain based on "idx" relative to the number of elements in `tarballs` ... but `collect_tb()` won't find the datasets we've already indexed in the loop, so `tarballs` will be the number _remaining_ from the original list (unless more have been added). So, for example, if we find 10, index 5, and then get a `SIGHUP`, our `idx` is 5, and `collect_tb()` returns the 5 we haven't indexed yet... so your message will report that we've indexed 5 and there are 0 remaining (5 - 5). I think this logic needs to be re-worked.

yeah I didn't think it thoroughly will work on it

> While the "update the list" is an intriguing idea, remember that the indexer will be triggered again in a short time, so I'm not sure how useful that really is. The consistency issues would go away if you dropped that and just continued to iterate through the original list.

What would be best?

Comment (riya-17) [reply lib/pbench/server/indexing_tarballs.py:442]: > @dbutenhof, perhaps a hyphen would help -- "re-collecting"? :-D
> 
> As for reassigning `tarballs`, see my comment below on `pbench-index.py` line 186.

yeah sure

File: server/bin/pbench-index.py
Comment (riya-17) [reply server/bin/pbench-index.py:115]: I will add that

File: server/bin/pbench-index.py
Comment (riya-17) [reply server/bin/pbench-index.py:186]: yeah, I have thought about the queue earlier but there were different reasons for not using that here. I will look into it again.

File: server/bin/pbench-index.py
Comment (riya-17) [reply server/bin/pbench-index.py:108]: yeah sure

File: server/bin/pbench-index.py
Comment (riya-17) [reply server/bin/pbench-index.py:115]: @dbutenhof which error should I report? like there is also a report generation at the end of the indexing that also reports no. of errors. So, should I just report the no. of errors the same ways or should I give details as well? also if there will be an error at the time of collection of tarballs inside sighup we can report that as well.

File: lib/pbench/server/indexing_tarballs.py
Comment (dbutenhof) [inline lib/pbench/server/indexing_tarballs.py:453]: Won't this add duplicates if we still haven't processed all the original tarballs?

**Review by webbnh — CHANGES_REQUESTED** [review]
It looks pretty good.  I've noted a few small things which should probably be fixed before merging.  The only real issue is updating the tarball list after a SIGHUP -- I think the list should be replaced, not augmented.

File: lib/pbench/server/indexing_tarballs.py
Comment (ndokos) [reply lib/pbench/server/indexing_tarballs.py:453]: We currently sort them by size: there have been instances of large tarballs taking "forever" to index, so dealing with the small ones first helped a bit. Of course, sorting this batch doesn't do anything about subsequent tarballs and even with SIGHUP adding new tarballs and re-sorting, that will not be solved: if a large tarball starts being indexed, everybody else (including the SIGHUP handling) will have to wait. OTOH, given a steady supply of tarballs and a twitchy SIGHUP finger, it's possible that  a large tarball will be starved. But no matter what we do with the current implementation, this will continue to be a problem. 

So I think, we hold the SIGHUP handling up to (or down to) the same standard: sort the current set of tarballs and process them in that order. Scheduling them in parallel, possilby with different queues handling different size classes, is something to think about. But we all know that the heuristics break down in particular circurmstances.

Scheduling is hard (I have to add that to my sig...)

Comment (ndokos) [reply lib/pbench/server/indexing_tarballs.py:453]: The only thing that we might want to check for sanity's sake is that the new list is a superset of what is left of the old list. I'm not sure that's particularly important  and I don't think we should stop if that is not the case, but having an error message listing any "lost" tarballs might be worth having, even though it should never trigger.

**Review by ndokos — COMMENTED** [review]
A couple of comment inline.

File: server/bin/pbench-index.py
Comment (riya-17) [reply server/bin/pbench-index.py]: oh, I must have missed the comment ;P

File: server/bin/pbench-index.py
Comment (riya-17) [reply server/bin/pbench-index.py:116]: ok sure

File: lib/pbench/server/indexing_tarballs.py
Comment (riya-17) [reply lib/pbench/server/indexing_tarballs.py:238]: yeah sure

File: lib/pbench/server/indexing_tarballs.py
Comment (riya-17) [reply lib/pbench/server/indexing_tarballs.py:144]: yeah this will be better

File: lib/pbench/server/indexing_tarballs.py
Comment (riya-17) [reply lib/pbench/server/indexing_tarballs.py:453]: oh, I must have forgotten about replacing the list it just gets out of my mind. Checking for the superset and logging the lost tarballs is a good idea I will add that.

File: lib/pbench/server/indexing_tarballs.py
Comment (riya-17) [reply lib/pbench/server/indexing_tarballs.py:460]: ok

**Comment by riya-17** [comment]
@ndokos @dbutenhof @webbnh I have addressed all the comments. Please take a look

File: lib/pbench/server/indexing_tarballs.py
Comment (dbutenhof) [inline lib/pbench/server/indexing_tarballs.py:449]: I don't expect this to happen, and I'm not sure we need another iteration of the PR to fix it, but a little more context might have been useful if this message ever _does_ appear. Some of the tarballs we thought we still had to process are no longer showing as `TO-INDEX`, and that's puzzling. Simply listing the "missing" tarballs makes this message a bit more mysterious than it needs to be.

File: lib/pbench/server/indexing_tarballs.py
Comment (dbutenhof) [inline lib/pbench/server/indexing_tarballs.py:233]: It occurs to me that `idx` isn't a particularly meaningful name anymore. It's a "count of processed tarballs". But while it might be nice to make this more clear should you make some other change, I don't want to push you to make another commit for this small detail.

File: server/bin/pbench-index.py
Comment (dbutenhof) [inline server/bin/pbench-index.py:184]: This isn't really a great use of `idx` either. Generally, short and abbreviated names should be used for things like "throwaway" loop variables, with longer and more meaningful names for variable of more importance and wider scope.

I suppose the fact that I'm paying attention to minor details on this review pass could be because I'm not seeing any important issues.

Again, not anything that needs to be resolved right now, but something to consider if you have to make another change for some other reason.

**Review by dbutenhof — APPROVED** [review]

File: lib/pbench/server/indexing_tarballs.py
Comment (riya-17) [reply lib/pbench/server/indexing_tarballs.py:449]: I was quite not sure about the log message I should put. So. I will just change that

File: lib/pbench/server/indexing_tarballs.py
Comment (webbnh) [inline lib/pbench/server/indexing_tarballs.py:444]: Here's an awkward question:  if `new_tb` evaluates to `False`, then we won't do the superset test and, more importantly, we won't update `tb_deque`.  The principal reason why `new_tb` would evaluate to `False` (accompanied by a "success" status) would be if someone went in and deleted all of the two-be-indexed datasets, presumably in an attempt to abort indexing.  So, what is going to happen next is that the tarball open (or something similar) will fail...which seems ugly.

This is a long-winded setup to ask, why are we testing `new_tb` at line 444?  Wouldn't it be better to just proceed based on the value of `status`?  (If `new_tb` turns out to be empty, won't proceeding make happen what we want to have happen?)

(Riya, I leave it up to you to decide how to address this -- you could fix it in this PR, do it in another, or just capture it as an issue.)

**Review by dbutenhof — APPROVED** [review]

**Review by webbnh — DISMISSED** [review]
Looks good.

Unfortunately, I found an issue...but maybe it doesn't need to be dealt with in _this_ PR.

File: lib/pbench/server/indexing_tarballs.py
Comment (riya-17) [reply lib/pbench/server/indexing_tarballs.py:444]: > Here's an awkward question: if `new_tb` evaluates to `False`, then we won't do the superset test and, more importantly, we won't update `tb_deque`. The principal reason why `new_tb` would evaluate to `False` (accompanied by a "success" status) would be if someone went in and deleted all of the two-be-indexed datasets, presumably in an attempt to abort indexing. So, what is going to happen next is that the tarball open (or something similar) will fail...which seems ugly.

@webbnh The reason that new_tb would evaluate as false is when there aren't any tarballs that need indexing("TO-INDEX" folder is empty). Maybe l.128 through l.130 can clear something up.


>


>
> This is a long-winded setup to ask, why are we testing `new_tb` at line 444? Wouldn't it be better to just proceed based on the value of `status`? (If `new_tb` turns out to be empty, won't proceeding make happen what we want to have happen?)

we could have proceeded just with status value but if there would have been no tarballs then all the checkings and replacing inside condition will all just not make sense.
if 'new_tb' will turn out to be empty then 'tb_deque' will run as it should and will end as soon as 'tb_deque' is empty.

Am I missing something?

Comment (webbnh) [reply lib/pbench/server/indexing_tarballs.py:444]: If we're running this code, then `tb_deque` had at least one entry in it -- let's assume that it had several -- and something sent a `SIGHUP` which made us call `collect_tb()`.  Now, for the sake of argument, suppose that the TO-INDEX folder is now empty (somehow...), which means that `new_tb` will be empty, but, for the sake of argument, suppose that `tb_deque` is _not_ empty (because we've indexed only the first item of the several).

At this point, we have an interesting choice.  We _could_ update `tb_deque` to be empty, which I think makes good sense, because there is presumably nothing in the TO-INDEX folder, now.  This will (I think) allow indexing to conclude in an orderly fashion.  The alternative, as currently coded, is to ignore the update, do nothing to `tb_deque`, and, on the next iteration of the loop, discover that the next tarball is nowhere to be found.

> Am I missing something?

Sleep?  :-D

Comment (webbnh) [reply lib/pbench/server/indexing_tarballs.py:444]: (Or, am _I_ missing something??  :-) )

Comment (riya-17) [reply lib/pbench/server/indexing_tarballs.py:444]: It is quite an exceptional condition but there is no harm in making it work. As, either both(tb_deque and new_tb) should be empty or both should be containing the same elements.

> Sleep? :-D

haha

Comment (riya-17) [reply lib/pbench/server/indexing_tarballs.py:444]: but yeah there is one more thing it will also add the somehow deleted tarballs in the missing log after not passing the superset condition.

**Review by dbutenhof — APPROVED** [review]

File: lib/pbench/server/indexing_tarballs.py
Comment (riya-17) [reply lib/pbench/server/indexing_tarballs.py]: fixed

**Review by webbnh — APPROVED** [review]

File: lib/pbench/server/indexing_tarballs.py
Comment (webbnh) [reply lib/pbench/server/indexing_tarballs.py]: > it will also add the somehow deleted tarballs in the missing log after not passing the superset condition.

And that's exactly what I think we want:  the tarballs were there when we started, and they are not there now, and they didn't get indexed, and that's weird and worth logging.

**Pull request merged**
